### PR TITLE
Feature/query capabilities

### PR DESCRIPTION
**Pull request by jtary**
Add query command for checking if a particular feature is supported.

connect to strongloop-internal/scrum-nodeops#453

### CODE REVIEW

File: lib/targetctl.js [hunk 1]
```diff
@@ -26,6 +26,19 @@ function getWorkerInfo() {
   };
 }
 
+function queryCapabilities(req, rsp, callback) {
+  var worker = getWorkerInfo();
+
+  rsp.cmd = 'query-capabilities';
```
Comment (sam-github) [inline lib/targetctl.js]: not necessary, the response is correlated with the request - it comes as a callback to the request() function,. `_appCommand` in your case.

File: lib/targetctl.js [hunk 2]
```diff
@@ -26,6 +26,19 @@ function getWorkerInfo() {
   };
 }
 
+function queryCapabilities(req, rsp, callback) {
+  var worker = getWorkerInfo();
```
Comment (sam-github) [inline lib/targetctl.js]: dead code?

File: lib/targetctl.js [hunk 3]
```diff
@@ -26,6 +26,19 @@ function getWorkerInfo() {
...
+
+  rsp.cmd = 'query-capabilities';
+  rsp.feature = req.feature;
```
Comment (sam-github) [inline lib/targetctl.js]: ditto for this

File: lib/targetctl.js [hunk 4]
```diff
@@ -26,6 +26,19 @@ function getWorkerInfo() {
...
+
+  if (rsp.feature === 'watchdog') {
+    rsp.supported = agent().internal.supports.watchdog;
```
Comment (sam-github) [inline lib/targetctl.js]: if its not supported, would be useful to have a .reasons, which would be an array of strings, each of which is a reason why its not supported, for example:

"watchdog not supported on this platform", and "addon not compiled", and "feature not licensed" (all three could be true at the same time). I can add the code to figure out all three, if you want.

**Comment by jtary** [comment]
@sam-github PTAL

File: lib/capabilities.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  var feature = capabilities[featureName];
+
+  if (feature == null) {
+    callback(false, ['unknown capability']);
+    return;
+  }
+
+  var reasons = [];
+  var status = true;
+  var count = feature.checks.length;
```
Comment (sam-github) [inline lib/capabilities.js]: I  think this is an [async.map](https://github.com/caolan/async#map)

File: lib/targetctl.js [hunk 6]
```diff
@@ -26,6 +27,32 @@ function getWorkerInfo() {
...
+  var features = req.feature;
+  var support = rsp.capabilities = {};
+
+  if (typeof features == 'string') {
+    features = features.split(',');
+  } else {
+    features = capabilities.list();
+  }
+
+  var count = features.length;
```
Comment (sam-github) [inline lib/targetctl.js]: sorry, this is the async.map

File: lib/capabilities.js (hunk 5 above)
Comment (sam-github) [inline lib/capabilities.js]: this would be an async.parallel

**Comment by sam-github** [comment]
The one check isn't async, and I can't think of any checks off the top of my head that would be async. Still, its not bad to assume it could be async, but it would be better to have less ad-hoc async code. Its good the reason array is allowed to be multi-line.

I'd add a comment about the expected form of the reasons: capitalized full sentences with periods, uncapitalized phrases? A mini-style guide would help, I think, when we add some extra reasons for watchdog and the other features.

File: lib/targetctl.js [hunk 7]
```diff
@@ -26,6 +28,34 @@ function getWorkerInfo() {
...
+      support[feature] = {
+        status: status,
+        reasons: reasons
+      };
+
+      callback(null, support[feature]);
+    });
+  }, function(err) {
+    if (err) {
+      rsp.error = err;
```
Comment (sam-github) [inline lib/targetctl.js]: `String(err)`, not the object.

Comment (sam-github) [inline lib/targetctl.js]: actually, `err.message`

**Comment by sam-github** [comment]
One comment, otherwise, LGTM.

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
Tests are failing, too, but it looks like they tested the wrong hash. I'll keep an eye on it.

**Comment by jtary** [comment]
Fix the message attachment, I'll keep an eye on the unit tests.

**Comment by sam-github** [comment]
LGTM, squash and merge.

**Pull request merged**